src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -477,9 +476,15 @@ public async Task ConnectAsync(CancellationToken cancellation = default(Cancella
         /// <param name="previous">Previous network state of the peer.</param>
         private async Task OnStateChangedAsync(NetworkPeerState previous)
         {
-            bool insideCallback = this.onDisconnectedAsyncContext.Value == null;
-            if (!insideCallback)
-                this.onDisconnectedAsyncContext.Value = new DisconnectedExecutionAsyncContext();
+            bool insideCallback;
+
+            lock (this.onDisconnectedAsyncContext)
```

Comment:
This PR is getting too big, from a simple change `insideCallback = this.onDisconnectedAsyncContext.Value != null; ` we change ValueLocal (this has special logic it will preserve the context of a task) and introduce locking.

Can you explain in more details why the two additional changes?

Reply:
@dangershony, As per @zeptin's comment above we want to avoid race conditions.  The value of the context variable was being tested and set without any locking across those two operations. This could lead to issues if both `OnStateChangedAsync` and `ProcessMessageAsync` try to conduct these operations simultaneously. This led to the introduction of locks.

Looking at the code at that point it was clear that `AsyncLocal` was not introducing any benefits so it was replaced with a similar structure that does not have the relatively opaque behavior of `AsyncLocal`.

Reply:
I think the PR should still be in two parts a PR to fix the `this.onDisconnectedAsyncContext.Value != null` bug on its own.

Reply:
`AsyncLocal` is an instance that lives in the context of the current executing Task (similar to ThreadLocal but for TPL) if there are locks does it make sense to keep the flag `insideCallback`?It Means one of the behaviours called in to the method form the current Task context and we know 
that becuse `onDisconnectedAsyncContext.Value` is null.

I am just not too sure this makes sense anymore, also I am not sure ther ecan be a race condition at all, what is the race on what resource?